libsolidity/analysis/DeclarationContainer.h

```diff
@@ -38,17 +40,26 @@ namespace solidity::frontend
 class DeclarationContainer
 {
 public:
+	using Homonyms = std::vector<std::pair<langutil::SourceLocation const*, Declaration const*>>;
+
 	explicit DeclarationContainer(
 		ASTNode const* _enclosingNode = nullptr,
-		DeclarationContainer const* _enclosingContainer = nullptr
+		DeclarationContainer* _enclosingContainer = nullptr
 	):
-		m_enclosingNode(_enclosingNode), m_enclosingContainer(_enclosingContainer) {}
+		m_enclosingNode(_enclosingNode), m_enclosingContainer(_enclosingContainer)
+	{
+		if (_enclosingContainer)
+			_enclosingContainer->m_innerContainers.emplace_back(this);
```

Comment:
This is fragile. We have to make sure that no `DeclarationContainer`s get destructed earlier.

For instance, [`DeclarationRegistrationHelper::enterNewSubScope`](https://github.com/ethereum/solidity/pull/9877/files#diff-653a1e23aa75715b13289b03e2f20947R612) was rewritten to avoid temporary containers.